src/system-module.adoc

```diff
@@ -0,0 +1,257 @@
+
+[{nrm}]
+----
+(make_symbol (content text\...)) \-> symbol
+----
+
+Like `make_string` but produces a symbol.
+
+
+===== `make_list`
+
+[{nrm}]
+----
+(make_list (vals any\...)) \-> list
+----
+
+Produces a non-null, unannotated list from any number of inputs.
+Template expressions of the form `[E~1~, …, E~n~]` are equivalent to `(make_list E~1~ … E~n~)`.
+
+
+===== `make_sexp`
+
+[{nrm}]
+----
+(make_sexp (vals any\...)) \-> sexp
+----
+
+Like make_list but produces a sexp.
+
+[{nrm}]
+----
+(:make_sexp)      ⇒ ()
+(:make_sexp null) ⇒ (null)
+----
+
+
+===== `make_struct`
+
+[{nrm}]
+----
+(make_struct (kv any\...)) \-> struct
+----
+
+Produces a non-null, unannotated struct from any number of elements.
+The ``kv``s are processed in order, incrementally adding fields to an initially-empty struct.
+Various forms of ``kv``s are allowed:
+
+* A (non-null) string or symbol is treated as a field name, and MUST be followed by another value
+to comprise a key-value pair in the result. Annotations on the field name are discarded.
+* A (non-null) struct is merged into the result as-is, after discarding annotations.
+* Any other type of value evokes an expansion error.
+
+Template expressions of the form `{T~1~:E~1~, …, T~n~:E~n~}` are equivalent to
+`(make_struct (*literal* T~1~) E~1~ … (*literal* T~n~) E~n~)`, assuming that no expression E
+produces more than one value.
+In that case, the `make_struct` variant would misbehave: the second value produced by E would be
+treated as the next key.
+
+
+===== `make_decimal`
+
+[{nrm}]
+----
+(make_decimal (coefficient int) (exponent int)) \-> decimal
+----
+
+Since decimal is already compact, this is perhaps most useful in conjunction with packed arrays.
+
+TODO if the coefficient were decimal, this could re-scale values. Useful?
+
+
+===== `make_float` ❓
+
+[{nrm}]
+----
+(make_float ieee) \-> float
+----
+
+Included for completeness, but of unclear utility.
+
+TODO coerce an int or decimal to float?  Perhaps useful to use fixed-width ints to encode various float widths? This may not be useable to convert “IEEE bits” to float, since they would be converted to int before arriving here.
+
+
+===== `make_timestamp`
+
+[{nrm}]
+----
+(make_timestamp
+  (year int) (month int) (day int?)
+  (hour int?)  (minute int?) (second decimal?)
+  (offset int?))
+  -> timestamp
+----
+Produces a non-null, unannotated timestamp at various levels of precision.
+When `offset` is absent, the result has unknown local offset; offset `0` denotes UTC.
+
+TODO Reconsider offset semantics, perhaps default should be UTC.
+
+Example:
+
+[{nrm}]
+----
+(*macro* ts_today
+  \((hour uint8) (minute uint8) (seconds_millis uint32))
+  (make_timestamp 2022 04 28 hour minute
+    (decimal seconds_millis -3) 0))
+----
+
+
+===== `annotate`
+
+[{nrm}]
+----
+(annotate (ann [text]*) value) \-> any
+----
+
+Produces the `value` prefixed with the annotations ``ann``s.
+Each `ann` must be a non-null, unannotated string or symbol.
```

Comment:
Here we require 'unannotated' while we ignore annotations elsewhere.

Reply:
Good call-out.  Now that you mention it, I feel icky about ignoring annotations.  That seems likely to lead to users of the macro being surprised by data-loss.

We could use a written-down design principle on this sort of thing, as well as related topics like handling of nulls (typed, untyped, and mis-typed).